DPGAnalysis/SiStripTools/plugins/EventWithHistoryProducerFromL1ABC.cc

```diff
@@ -104,16 +107,26 @@ void EventWithHistoryProducerFromL1ABC::produce(edm::Event& iEvent, const edm::E
   } else {
     Handle<L1AcceptBunchCrossingCollection> pIn;
     iEvent.getByToken(_l1abccollectionToken, pIn);
+    Handle<TCDSRecord> tcds_pIn;
+    iEvent.getByToken(_tcdsRecordToken, tcds_pIn);
+    const auto& tcdsRecord = *tcds_pIn.product();
+    bool useTCDS = (tcds_pIn.product());
 
     // offset computation
 
     long long orbitoffset = 0;
     int bxoffset = 0;
-    if (!_forceNoOffset) {
-      for (L1AcceptBunchCrossingCollection::const_iterator l1abc = pIn->begin(); l1abc != pIn->end(); ++l1abc) {
-        if (l1abc->l1AcceptOffset() == 0) {
-          orbitoffset = (long long)l1abc->orbitNumber() - (long long)iEvent.orbitNumber();
-          bxoffset = l1abc->bunchCrossing() - iEvent.bunchCrossing();
+
+    if (useTCDS) {
+      orbitoffset = (long long)tcdsRecord.getOrbitNr() - (long long)iEvent.orbitNumber();
+      bxoffset = tcdsRecord.getBXID() - iEvent.bunchCrossing();
+    } else {
+      if (!_forceNoOffset) {
+        for (L1AcceptBunchCrossingCollection::const_iterator l1abc = pIn->begin(); l1abc != pIn->end(); ++l1abc) {
+          if (l1abc->l1AcceptOffset() == 0) {
+            orbitoffset = (long long)l1abc->orbitNumber() - (long long)iEvent.orbitNumber();
+            bxoffset = l1abc->bunchCrossing() - iEvent.bunchCrossing();
+          }
```

Comment:
@vischia I think we need also a new `EventWithHistory` constructor.
The one called below at line 134 in this file:
```C++
std::unique_ptr<EventWithHistory> pOut(new EventWithHistory(iEvent, *pIn, orbitoffset, bxoffset));
```
uses `Handle<L1AcceptBunchCrossingCollection> pIn`.
Now IIUC this object is put in the event by the SCAL unpacker here:
https://github.com/cms-sw/cmssw/blob/master/EventFilter/ScalersRawToDigi/src/ScalersRawToDigi.cc#L55.
This means that if SCAL is not in the DAQ the object used here will be empty.

Reply:
Oh, I see. Thanks for spotting @mmusich , I'll try to fix this on Monday.

Reply:
Hi @mmusich , I am adding the constructor, but I can't find the equivalent in https://github.com/cms-sw/cmssw/blob/master/DataFormats/TCDS/interface/TCDSRecord.h of the l1acceptOffset from https://github.com/cms-sw/cmssw/blob/master/DataFormats/Scalers/interface/L1AcceptBunchCrossing.h .

While in the other modules I solved by just removing the check ```if l1abc->l1AcceptOffset() == 0)```, https://github.com/cms-sw/cmssw/blob/master/DPGAnalysis/SiStripTools/src/EventWithHistory.cc#L37 requires to compute a quantity and use it later, so it's not clear to me how to replace it with something from TCDSRecord

Reply:
Finding out how to deal with it is part of the ticket mandate.

Reply:
I haven't checked the details, but maybe this https://github.com/cms-sw/cmssw/blob/master/DataFormats/TCDS/interface/TCDSRecord.h#L109 can help?